src/core/project/qgsproject.cpp

```diff
@@ -1324,63 +1324,117 @@ static  QgsMapLayer::ReadFlags projectFlagsToLayerReadFlags( Qgis::ProjectReadFl
 
-    dataSource = QgsProviderRegistry::instance()->relativeToAbsoluteUri( provider, dataSource, context );
+    layerToLoad.dataSource = QgsProviderRegistry::instance()->relativeToAbsoluteUri( layerToLoad.provider, layerToLoad.dataSource, context );
 
-    QgsDataProvider::ProviderOptions options( {context.transformContext()} );
-    QgsDataProvider::ReadFlags flags = QgsMapLayer::providerReadFlags( node, layerReadFlags );
+    layerToLoad.options = QgsDataProvider::ProviderOptions( {context.transformContext()} );
+    layerToLoad.flags = QgsMapLayer::providerReadFlags( node, layerReadFlags );
 
     // Requesting credential from worker thread could lead to deadlocks because the main thread is waiting for worker thread to fininsh
-    flags.setFlag( QgsDataProvider::SkipCredentialsRequest, true );
+    layerToLoad.flags.setFlag( QgsDataProvider::SkipCredentialsRequest, true );
 
-    QgsRunnableProviderCreator *run = new QgsRunnableProviderCreator( layerId, provider, dataSource, options, flags );
-    runnables.append( run );
-    QObject::connect( run, &QgsRunnableProviderCreator::providerCreated, this, [&]( bool isValid )
+    layersToLoad.insert( layerToLoad.layerId, layerToLoad );
+  }
+
+  while ( !layersToLoad.isEmpty() )
+  {
+    const QList<LayerToLoad> layersToAttemptInParallel = layersToLoad.values();
+    QString layerToAttemptInMainThread;
+
+    QVector<QgsRunnableProviderCreator *> runnables;
+    QThreadPool threadPool;
+    threadPool.setMaxThreadCount( QgsSettingsRegistryCore::settingsLayerParallelLoadingMaxCount->value() );
+
+    for ( const LayerToLoad &lay : layersToAttemptInParallel )
     {
+      QgsRunnableProviderCreator *run = new QgsRunnableProviderCreator( lay.layerId, lay.provider, lay.dataSource, lay.options, lay.flags );
+      runnables.append( run );
+      QObject::connect( run, &QgsRunnableProviderCreator::providerCreated, run, [&]( bool isValid, const QString & layId )
+      {
+        if ( isValid )
+        {
+          layersToLoad.remove( layId );
+          i++;
+          emit layerLoaded( i, totalProviderCount );
+        }
+        else
+        {
+          if ( layerToAttemptInMainThread.isEmpty() )
+            layerToAttemptInMainThread = layId;
+          threadPool.clear(); //we have to stop all loading provider to try this layer in main thread and maybe have credentials
+        }
+
+        if ( i == parallelLayerNodes.count() || !isValid )
+          loop.quit();
+      } );
+      threadPool.start( run );
+    }
+    if ( !parallelLayerNodes.isEmpty() )
+      loop.exec();
+
+    threadPool.waitForDone(); // to be sure all threads are finished
+
+    // First we take all valid providers
+    for ( QgsRunnableProviderCreator *run : std::as_const( runnables ) )
+    {
+      std::unique_ptr<QgsDataProvider> provider( run->dataProvider() );
+      if ( provider && provider->isValid() )
+        loadedProviders.insert( run->layerId(), provider.release() );
+    }
```

Comment:
Can we insert to the `loadedProviders` directly in the slot for `QgsRunnableProviderCreator::providerCreated` when `isValid` is true? That would make the code a bit easier to understand, as this for-loop would not be needed...?

Reply:
let's try.